libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,584 @@
+    while (++i <= _last_index) { // never removes the first item. This should always be {0,0,0}
+        if (_path[i] != Vector3f(0.0f, 0.0f, 0.0f)) {
+            _path[++j] = _path[i];
+        } else {
+            removed++;
+        }
+    }
+    _last_index -= removed;
+}
+
+/**
+*  Returns the closest distance in 3D space between any part of two input segments, defined from p1 to p2 and from p3 to p4.
+*  Also returns the point which is halfway between
+*
+*  Limitation: This function does not work for parallel lines. In this case, it will return FLT_MAX. This does not matter for the path cleanup algorithm because
+*  the pruning will still occur fine between the first parallel segment and a segment which is directly before or after the second segment.
+*/
+AP_SafeRTL::dist_point AP_SafeRTL::segment_segment_dist(const Vector3f &p1, const Vector3f &p2, const Vector3f &p3, const Vector3f &p4)
+{
+    Vector3f line1 = p2-p1;
+    Vector3f line2 = p4-p3;
+    Vector3f line_start_diff = p1-p3; // from the beginning of the second line to the beginning of the first line
+
+    // these don't really have a physical representation. They're only here to break up the longer formulas below.
+    float a = line1*line1;
+    float b = line1*line2;
+    float c = line2*line2;
+    float d = line1*line_start_diff;
+    float e = line2*line_start_diff;
+
+    // the parameter for the position on line1 and line2 which define the closest points.
+    float t1 = 0.0f;
+    float t2 = 0.0f;
+
+    // if lines are almost parallel, return a garbage answer. This is irrelevant, since the loop
+    // could always be pruned start/end of the previous/subsequent line segment
+    if (is_zero((a*c)-(b*b))) {
+        return {FLT_MAX, Vector3f(0.0f, 0.0f, 0.0f)};
+    } else {
+        t1 = (b*e-c*d)/(a*c-b*b);
+        t2 = (a*e-b*d)/(a*c-b*b);
+
+        // restrict both parameters between 0 and 1.
+        t1 = constrain_float(t1, 0.0f, 1.0f);
+        t2 = constrain_float(t2, 0.0f, 1.0f);
+
+        // difference between two closest points
+        Vector3f dP = line_start_diff+line1*t1-line2*t2;
+
+        Vector3f halfway_point = (p1+line1*t1 + p3+line2*t2)/2.0f;
+        return {dP.length(), halfway_point};
+    }
+}
+
+
+/**
+*  Returns the closest distance from a point to a 3D line. The line is defined by any 2 points
+*/
+float AP_SafeRTL::point_line_dist(const Vector3f &point, const Vector3f &line1, const Vector3f &line2)
+{
+    // triangle side lengths
+    float a = HYPOT(point, line1);
+    float b = HYPOT(line1, line2);
+    float c = HYPOT(line2, point);
+
+    // semiperimeter of triangle
+    float s = (a+b+c)/2.0f;
+
+    float area_squared = s*(s-a)*(s-b)*(s-c);
+    // must be constrained above 0 because a triangle where all 3 points could be on a line. float rounding could push this under 0.
+    if (area_squared < 0.0f) {
+        area_squared = 0.0f;
+    }
+    float area = safe_sqrt(area_squared);
+    return 2.0f*area/b;
```

Comment:
need to handle the case where b is 0 ... it shouldn't happend but with float precision and weird people it could !